src/Datadog.Trace/Containers/ContainerInfo.cs

```diff
@@ -0,0 +1,118 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text.RegularExpressions;
+using System.Threading;
+
+namespace Datadog.Trace.Containers
+{
+    /// <summary>
+    /// Utility class with methods to interact with container hosts.
+    /// </summary>
+    internal static class ContainerInfo
+    {
+        private const string ControlGroupsFilePath = "/proc/self/cgroup";
+
+        private const string ContainerIdRegex = @"^(?:\d+):(?:[^:]*):/?(?:.+/)([0-9a-f]{8}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{12}|[0-9a-f]{64}(?:\.scope)?)$";
+
+        private static readonly Lazy<string> ContainerId = new Lazy<string>(GetContainerIdInternal, LazyThreadSafetyMode.ExecutionAndPublication);
+
+        /// <summary>
+        /// Gets the id of the container executing the code.
+        /// Return <c>null</c> if code is not executing inside a supported container.
+        /// </summary>
+        /// <returns>The container id or <c>null</c>.</returns>
+        public static string GetContainerId()
+        {
+            return ContainerId.Value;
+        }
+
+        /// <summary>
+        /// Extract the container id from the specified cgroup file contents.
+        /// </summary>
+        /// <remarks>
+        /// This method is used when passing the entire cgroup text for testing purposes.
+        /// When reading from file system, we only read one line at a time until we find a match,
+        /// which is usually the first line.
+        /// </remarks>
+        /// <param name="contents">The contents of a cgroup file.</param>
+        /// <returns>The container id if a match is found; otherwise, <c>null</c>.</returns>
+        public static string ParseCgroupText(string contents)
```

Comment:
Very minor thing, but I would prefer `ParseCgroupText` and `SplitLines` go into the test assembly instead of here so we can limit the test's method calls to the interface we'll be using in `Datadog.Trace`.

Reply:
I like this. Keeps this file leaner. Changed in a2b39232f6c53f53fff1bf2a95930b2a06758f69.